encrypted-config-value/src/main/java/com/palantir/config/crypto/value/EncryptedValue.java

```diff
@@ -0,0 +1,115 @@
+import com.google.common.base.Throwables;
+import com.google.common.io.BaseEncoding;
+import com.palantir.config.crypto.KeyPair;
+import com.palantir.config.crypto.KeyWithAlgorithm;
+import com.palantir.config.crypto.algorithm.Algorithm;
+import com.palantir.config.crypto.algorithm.Algorithms;
+import java.io.IOException;
+
+@JsonSubTypes({
+        @JsonSubTypes.Type(value = AesEncryptedValue.class),
+        @JsonSubTypes.Type(value = RsaEncryptedValue.class)
+        })
+@JsonTypeInfo(property = "type", use = JsonTypeInfo.Id.NAME)
+public abstract class EncryptedValue {
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+    private static final String PREFIX = "enc:";
+
+    public abstract <T> T accept(EncryptedValueVisitor<T> visitor);
+
+    public static boolean isEncryptedValue(String value) {
+        return value.startsWith(PREFIX);
+    }
+
+    public static EncryptedValue deserialize(String value) {
+        if (!value.startsWith(PREFIX)) {
+            throw new IllegalArgumentException("Missing enc: prefix: " + value);
+        }
+
+        String suffix = value.substring(PREFIX.length());
+        byte[] bytes = BaseEncoding.base64().decode(suffix);
+
+        // this is a bit dubious but hopefully we can remove the legacy stuff soonish
+        try {
+            return MAPPER.readValue(bytes, EncryptedValue.class);
+        } catch (IOException e) {
+            return ImmutableLegacyEncryptedValue.of(bytes);
+        }
+    }
+
+    @JsonIgnore
+    public final String serialize() {
+        byte[] bytes = accept(new EncryptedValueVisitor<byte[]>() {
+            @Override
+            public byte[] visit(LegacyEncryptedValue legacyEncryptedValue) {
+                return legacyEncryptedValue.getCiphertext();
+            }
+
+            @Override
+            public byte[] visit(AesEncryptedValue aesEncryptedValue) {
+                try {
+                    return MAPPER.writeValueAsBytes(aesEncryptedValue);
+                } catch (JsonProcessingException e) {
+                    throw Throwables.propagate(e);
+                }
+            }
+
+            @Override
+            public byte[] visit(RsaEncryptedValue rsaEncryptedValue) {
+                try {
+                    return MAPPER.writeValueAsBytes(rsaEncryptedValue);
+                } catch (JsonProcessingException e) {
+                    throw Throwables.propagate(e);
+                }
+            }
+        });
+
+        return PREFIX + BaseEncoding.base64().encode(bytes);
+    }
+
+    @JsonIgnore
+    public final String getDecryptedValue(KeyWithAlgorithm kwa) {
+        Algorithm algorithm = Algorithms.getInstance(kwa.getAlgorithm());
+        return algorithm.getDecryptedString(this, kwa);
+    }
+
+    /**
+     * Tries to decrypt using the key at the default path.
+     * @see KeyPair#DEFAULT_PUBLIC_KEY_PATH
+     * @return the decrypted value
+     */
+    @JsonIgnore
+    public final String getDecryptedValue() {
```

Comment:
Yeah, this method in particular feels kind of weird and overly magical. It also means we're reading the key off disk and parsing it for every encrypted value in a config.